test/cpp/end2end/concurrent_resumption_test.cc

```diff
@@ -0,0 +1,143 @@
+#include <grpcpp/server.h>
+#include <grpcpp/server_builder.h>
+
+#include "src/core/lib/iomgr/load_file.h"
+// #include "src/core/lib/security/authorization/audit_logging.h"
+// #include
+// "src/core/lib/security/authorization/grpc_authorization_policy_provider.h"
+// #include "src/core/lib/security/credentials/fake/fake_credentials.h"
+// #include "src/cpp/client/secure_credentials.h"
+// #include "src/cpp/server/secure_server_credentials.h"
+// #include "src/proto/grpc/testing/echo.grpc.pb.h"
+// #include "test/core/util/audit_logging_utils.h"
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/core/util/port.h"
+#include "test/core/util/test_config.h"
+#include "test/core/util/tls_utils.h"
+#include "test/cpp/end2end/test_service_impl.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+
+constexpr char kCaCertPath[] = "src/core/tsi/test_creds/ca.pem";
+constexpr char kServerCertPath[] = "src/core/tsi/test_creds/server1.pem";
+constexpr char kServerKeyPath[] = "src/core/tsi/test_creds/server1.key";
+constexpr char kClientCertPath[] = "src/core/tsi/test_creds/client.pem";
+constexpr char kClientKeyPath[] = "src/core/tsi/test_creds/client.key";
+constexpr char kMessage[] = "Hello";
+
+std::string ReadFile(const char* file_path) {
+  grpc_slice slice;
+  GPR_ASSERT(
+      GRPC_LOG_IF_ERROR("load_file", grpc_load_file(file_path, 0, &slice)));
+  std::string file_contents(grpc_core::StringViewFromSlice(slice));
+  grpc_slice_unref(slice);
+  return file_contents;
+}
+
+class GrpcResumptionTest : public ::testing::Test {
+ protected:
+  void RunServer() {
+    std::string root_cert = ReadFile(kCaCertPath);
+    grpc::SslServerCredentialsOptions::PemKeyCertPair key_cert_pair = {
+        ReadFile(kServerKeyPath), ReadFile(kServerCertPath)};
+    grpc::SslServerCredentialsOptions sslOpts;
+    sslOpts.pem_key_cert_pairs.push_back(key_cert_pair);
+    sslOpts.pem_root_certs = root_cert;
+    sslOpts.force_client_auth = true;
+
+    grpc::ServerBuilder builder;
+    TestServiceImpl service_;
+
+    builder.AddListeningPort(server_addr_, grpc::SslServerCredentials(sslOpts));
+    builder.RegisterService("foo.test.google.fr", &service_);
+    server_ = builder.BuildAndStart();
+    std::cout << "GREG: server waiting on " << server_addr_ << "\n";
+    server_->Wait();
+  }
+
+  //   void TearDown() override {
+  //     if (server_ != nullptr) {
+  //       server_->Shutdown();
+  //       server_thread_->join();
+  //     }
+  //     server_thread_ = nullptr;
+  //   }
+
+  TestServiceImpl service_;
+  std::unique_ptr<Server> server_;
+  std::thread* server_thread_;
+  std::string server_addr_;
+};
+
+TEST_F(GrpcResumptionTest, ConcurrentResumption) {
+  int port = grpc_pick_unused_port_or_die();
+  std::ostringstream addr;
+  addr << "localhost:" << port;
+  server_addr_ = addr.str();
+  std::cout << "GREG: Serving on " << server_addr_ << "\n";
+  server_thread_ = new std::thread([&]() { RunServer(); });
```

Comment:
To avoid some flakiness I noticed while playing with the test, you'll want to use an `absl::Notification` instead of `sleep(1)`. You can do this as follows:

```
#include "absl/synchronization/notification.h"

void RunServer(absl::Notification* notification) {
 ...
 server_ = builder.BuildAndStart();
 notification->Notify();
 server_->Wait();
}

TEST_F(GrpcResumptionTest, ConcurrentResumption) {
   ...
   absl::Notification notification;
   server_thread_ = new std::thread([this, &notification]() { RunServer(&notification); });
   notification.WaitForNotification();
   ...
}
```

You'll also need to add `"absl/synchronization"` to the `external_deps` of the `grpc_cc_test`.

Reply:
Will definitely add that, the `sleep(1)` was my temporary solution while working on figuring on why the test wasn't working